File: ProgrammingBasics/30-Lenses-master/JavaScript/3-set.js
"use strict";

const getter = (prop) => (obj) => obj[prop];
const setter = (prop) => (val, obj) => ({ ...obj, [prop]: val });

const set = (lens, val, obj) => lens.set(val, obj);

const lens = (getter, setter) => ({
  get: (obj) => getter(obj),
  set: (val, obj) => setter(val, obj),
});

// Usage

const person = {
  name: "Marcus Aurelius",
  city: "Rome",
  born: 121,
};

const nameLens = lens(getter("name"), setter("name"));

console.log("set name:", set(nameLens, "Marcus", person));

File: ProgrammingBasics/30-Lenses-master/Haskell/3-set.hs
import Text.Printf

data Person = Person
  { _name :: String
  , _city :: String
  , _born :: Int
  } deriving Show
person = Person "Marcus Aurelius" "Rome" 121

data Lens s a = Lens (s -> a) (a -> s -> s)
nameGetter = _name
nameSetter name person = person { _name = name }
nameLens = Lens nameGetter nameSetter
set (Lens _ s) = s

main = printf "set name: %v\n" $ show $ set nameLens "Marcus" person
